[inductor] don't allow triton config pre_hook

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134633

The caching autotuner caches triton configs, and it doesn't try to hash or save the pre_hook from the config if it exists. If we had a config that had a pre_hook, then we might autotune -> save the config (without the pre_config) -> later load the saved config and try to run it, but this time without the pre_hook.

So this PR adds an assert and deletes the pre_hook handling. We can be confident that we didn't have functional pre_hooks, because the pre_hook handling tries to use `self.arg_name`, which doesn't exist.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

davidberard98 committed:
[inductor] don't allow triton config pre_hook

The caching autotuner caches triton configs, and it doesn't try to hash or save the pre_hook from the config if it exists. If we had a config that had a pre_hook, then we might autotune -> save the config (without the pre_config) -> later load the saved config and try to run it, but this time without the pre_hook.

So this PR adds an assert and deletes the pre_hook handling. We can be confident that we didn't have functional pre_hooks, because the pre_hook handling tries to use `self.arg_name`, which doesn't exist.

[ghstack-poisoned]

davidberard98 committed:
Update on "[inductor] don't allow triton config pre_hook"


The caching autotuner caches triton configs, and it doesn't try to hash or save the pre_hook from the config if it exists. If we had a config that had a pre_hook, then we might autotune -> save the config (without the pre_config) -> later load the saved config and try to run it, but this time without the pre_hook.

So this PR adds an assert and deletes the pre_hook handling. We can be confident that we didn't have functional pre_hooks, because the pre_hook handling tries to use `self.arg_name`, which doesn't exist.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

davidberard98 committed:
Update on "[inductor] don't allow triton config pre_hook"


The caching autotuner caches triton configs, and it doesn't try to hash or save the pre_hook from the config if it exists. If we had a config that had a pre_hook, then we might autotune -> save the config (without the pre_config) -> later load the saved config and try to run it, but this time without the pre_hook.

So this PR adds an assert and deletes the pre_hook handling. We can be confident that we didn't have functional pre_hooks, because the pre_hook handling tries to use `self.arg_name`, which doesn't exist.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

davidberard98 commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -182,6 +183,9 @@ def __init__(
 
         assert len(configs) > 0, "Non-empty TritonConfig list required for compiling"
+        # makes sure there are no pre-hooks on any of the triton configs
+        (validate_triton_config(cfg) for cfg in configs)
```
not sure if there's a better place to put the check? I chose here because the CachingAutotuner is the place where the triton runtime behavior is modeled (and where the pre_hook behavior is being omitted now), but open to suggestions if there's another place that's more obvious

shunting314 reviewed (commented):
Looks safe to remove. I vaguely remember @Jokeren uses these for profiling?

Jokeren commented:
> Looks safe to remove. I vaguely remember @Jokeren uses these for profiling?

If launcher is an autotuner, then yes.

Jokeren commented:
Feel free to merge even if it will cause any conflicts, there might be multiple problems that prevents profiling for torch compiled triton code. I plan to take a look soon and sort them out.

davidberard98 commented:
@Jokeren thanks for the context. I imagine it might be worth adding support for pre_hooks back in at some point, but I'm pretty sure the pre-hook handling here was broken anyway, so we should probably revisit this in the future if we do find cases where we need pre_hooks

davidberard98 commented:
@pytorchbot merge

davidberard98 commented:
@pytorchbot merge

jansel commented on torch/_inductor/runtime/runtime_utils.py:
```diff
@@ -65,6 +65,15 @@ def triton_config_to_hashable(cfg):
+    # right now, if a pre-hook is attached to the config, it will not be saved;
+    # and then it won't be used when the config is loaded from cache.
+    # So we assert - if we do get a pre_hook, it might get ignored after caching.
+    assert not hasattr(cfg, "pre_hook") or cfg.pre_hook is None
```
```suggestion
    assert getattr(cfg, "pre_hook", None) is None
```

jansel commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -182,6 +183,9 @@ def __init__(
 
         assert len(configs) > 0, "Non-empty TritonConfig list required for compiling"
+        # makes sure there are no pre-hooks on any of the triton configs
+        (validate_triton_config(cfg) for cfg in configs)
```
I don't think this will actually do anything since it is a unused generator:
```
>>> (print(x) for x in range(10))
<generator object <genexpr> at 0x7e9bd143e110>
>>> [print(x) for x in range(10)]
0
1
2
3
4
5
6
7
8
9
[None, None, None, None, None, None, None, None, None, None]
>>> 
```

davidberard98 replied:
thanks for catching!

jansel reviewed (changes requested):
Add test for this.

davidberard98 committed:
Update on "[inductor] don't allow triton config pre_hook"


The caching autotuner caches triton configs, and it doesn't try to hash or save the pre_hook from the config if it exists. If we had a config that had a pre_hook, then we might autotune -> save the config (without the pre_config) -> later load the saved config and try to run it, but this time without the pre_hook.

So this PR adds an assert and deletes the pre_hook handling. We can be confident that we didn't have functional pre_hooks, because the pre_hook handling tries to use `self.arg_name`, which doesn't exist.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

davidberard98 committed:
Update on "[inductor] don't allow triton config pre_hook"


The caching autotuner caches triton configs, and it doesn't try to hash or save the pre_hook from the config if it exists. If we had a config that had a pre_hook, then we might autotune -> save the config (without the pre_config) -> later load the saved config and try to run it, but this time without the pre_hook.

So this PR adds an assert and deletes the pre_hook handling. We can be confident that we didn't have functional pre_hooks, because the pre_hook handling tries to use `self.arg_name`, which doesn't exist.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot merge
